### ISSUE TITLE
Book won't download

### ISSUE THREAD

**Issue by anesone** [rmcrackan/Libation#350]
The only book I haven't been able to liberate is War and Peace.  It shows as an error when I try to liberate it. On Audible's site, it gives the option to download the full file or in parts. I have downloaded the full aax file,
It's there a way I can start the liberation process using the downloaded file?

**Comment by Mbucari** [comment]
Try deleting the files in your temporary download
%temp%\Libation

**Comment by anesone** [comment]
I just did that, and restarted the app.
I still get an error message: ERROR 000307
It doesn't show up in either of the folders under %temp%\Libation.

Any other ideas?

**Comment by Mbucari** [comment]
I can't do anything else without you posting your log file.

**Comment by anesone** [comment]
[Log202208_005.log](https://github.com/rmcrackan/Libation/files/9305896/Log202208_005.log)

**Comment by rmcrackan** [comment]
Thank you for your log. The error in it looks like you don't have permission. Are you currently able to download this book on your phone?

**Comment by rmcrackan** [comment]
@Mbucari whether or not it's this user's problem, I've gotten 2 common bug reports via reddit PMs. I'll create a new issue for it

**Comment by anesone** [comment]
I downloaded it from the website, and I just started the download via the
app on my phone with no problems.

**Comment by Mbucari** [comment]
@rmcrackan It looks like Audible changed the ASIN for that title.

The War and Peace asin Libation is requesting is B002V1PM18, but I can't find it on audible's website.

@anesone Can you please post an audible.com link to the copy of War and Peace you're trying to download?

**Comment by rmcrackan** [comment]
> B002V1PM18, but I can't find it on audible's website.

Not unusual. Public domain classics have high churn unless it's a popular recording. I have tons of old books like this.

What I think is interesting is that it's a licenserequest *failure*, not a rejection: `Unable to retrieve asset details from Sable(CDN), fallback to CDS`

**Comment by anesone** [comment]
This is the link associated to the book in my library: https://www.audible.com/pd/War-and-Peace-Audiobook/B002V1PM18?ref=a_library_t_c5_libItem_&pf_rd_p=80765e81-b10a-4f33-b1d3-ffb87793d047&pf_rd_r=07ZG5KFAR7SJK5ZE96BE

When I follow it from my library, It shows "Sorry, it looks like this title is no longer available."

This is the cover art link: https://m.media-amazon.com/images/I/51YzMzXm2eL._SL500_.jpg

This is the webplayer link: https://www.audible.com/webplayer?asin=B002V1PM18

This is the download link: https://www.audible.com/library/download?asin=B002V1PM18&codec=AAX

**Comment by rmcrackan** [comment]
Now that it's been a few days since the original error, can you try again and reply with your new log? I'd like to see if either or it works now. Or if not: if the error is the same.

As I mentioned before, it's curious that it's a failure and not a rejection. That is, when Libation asked audible for the book, it didn't say 'no', it said 'I can't'. This is sometimes a symptom of an issue on their end which resolves itself.

**Comment by anesone** [comment]
[Log202208_005.log](https://github.com/rmcrackan/Libation/files/9331562/Log202208_005.log)
Here is the log from my attempt this morning.

**Comment by rmcrackan** [comment]
Thanks. It's still `Unable to retrieve asset details from Sable(CDN), fallback to CDS`

@mkb79 I've got another weird one for you. Have you ever seen where a title is unable to be resolved with the API -- resulting in this message? The user can retrieve it with their phone. I assume their phone knows how to "fallback to CDS" but I don't.

**Comment by mkb79** [comment]
@rmcrackan 
I've got this error message only when downloading books in aax format. This occurs in most cases when downloading Audible Plus content.

**Comment by mkb79** [comment]
Take a look at [this](https://github.com/rmcrackan/Libation/issues/122) issue.

**Comment by rmcrackan** [comment]
@mkb79 Thanks. I'd forgotten about that one. In this case though, it's not a plus title -- it's just a regular book.

**Comment by mkb79** [comment]
@rmcrackan Do you downloading this title in aax or aaxc format?

**Comment by rmcrackan** [comment]
@mkb79 Neither format yet. It doesn't get that far. The error is happening for licenserequest

```
Request: https://api.audible.com/1.0/content/B002V1PM18/licenserequest

Response:
{
  "error_code": "000307",
  "message": "Unable to retrieve asset details from Sable(CDN), fallback to CDS for customerId:[[hidden]], marketplaceId:AF2M0KC94RCEA, asin:B002V1PM18, acr:null, skuLite:BK_BKOT_000096, aaaClientId:ApolloEnv:AudibleApiExternalRouterService/NA/ACLS/Prod"
}
```

**Comment by mkb79** [comment]
@rmcrackan 
This is a MultipartBook. Can you try to download the child asin B002V1ORNM. This licenserequest works for me but will give `License not granted`. The parent asin `B002V1PM18` gives me the same 404 error message (NotFound)!

**Comment by Mbucari** [comment]
@mkb79 You're right! I think this is the first time I've seen a multipart-only book in the wild.

@rmcrackan Call GetCatalogProductAsync on `B002V1PM18`. Its `ContentDeliveryType` is "MultiPartBook". We're going to have to re-write the backup routines to fix this...

**Comment by rmcrackan** [comment]
@mkb79 , @Mbucari Much thanks guys!! I should have a fix ready soon.

**Comment by rmcrackan** [comment]
Code is [complete](https://github.com/rmcrackan/Libation/commit/1510a86579a7bb19a8d372d355674955c2b5668e). I'll let you know when a new version goes out. I'm currently fighting through a publishing issue first.

**Comment by Mbucari** [comment]
@rmcrackan Something is not right here. I've found other titles whose ContentDeliveryType is MultiPartBook but that download as a single file without issue. e.g. `B00DCD0OXU`

When I try to download `B00DCD0OXU` using the new part downloading, I'm denied a license to the parts (the same thing which happened to @mkb79 when he requested `B002V1ORNM`)

I compared the `GetCatalogProductAsync` responses for both `B00DCD0OXU` and `B002V1PM18`, and I can't find any differences that could account for the latter resulting in an API error.

My new theory: `B002V1PM18` does not exist as an AAXC file and must be downloaded using the old `/FionaCDEServiceEngine` endpoint. `B002V1PM18` is a legacy recording from 2002 and can no longer be browsed on Audible.com, so perhaps they never bothered making an aaxc version of it. Even if that is true, I don't know how to tell the difference between books that need to be downloaded using `/1.0/content/{asin}/licenserequest` and those needing to use `/FionaCDEServiceEngine`. If there is a metadata clue, it's not returned by `GetCatalogProductAsync`.

**Comment by mkb79** [comment]
@Mbucari 
I could download Multipartbooks as a single file since now without any issue except the book above. As I inspected the asin via the catalog/product endpoint I found at that this is a Multipartbook too. I tried to download the child asin and these seams to work for THIS book. I saw that the book is from 2002, too. So your theory is absolutely possible. But I also don't know how to distinguish these books.

**Comment by rmcrackan** [comment]
Nevermind. I reverted that change

**Comment by Mbucari** [comment]
@anesone I think we're at an impasse. Without being able to debug this problem using your copy of War and Peace, I don't think we'll be able to find the problem.

I _may_ be able to solve this problem if I could access that copy of War and Peace, but the only way I can do that is through the account of someone who has already purchased it. If you're willing to trust me, you can email your AccountSettings.json file to me at mbucari1@gmail.com. That will allow me to access your audible library so I may attempt to download War and Peace and see what's going wrong. I completely understand if you don't want to do that, but it the last option I can think of to try and fix this.

**Comment by rmcrackan** [comment]
@anesone I'm going to close this. Not because it's fixed; because we've taken it as far as is possible without access to this very old book which is no longer in the general catalog. If you choose to collaborate with @Mbucari then the conversation can continue there. If if you don't, then this issue is at its end anyway.

I feel comfortable speaking for both of us when I say that with enough access, we're happy to help you continue looking into this further. If you are in any way uncomfortable with granting him this access then please do not do so. Your privacy is worth more than an audiobook.

**Comment by anesone** [comment]
Any thoughts on my other question in my first post?  Is there a way to convert a downloaded .aax to create the m4b, cue, jpg, etc?

**Comment by rmcrackan** [comment]
> Is there a way to convert a downloaded .aax to create the m4b, cue, jpg, etc?

[inAudible](https://github.com/rmcrackan/inAudible) was a fantastic program when it came out but it doesn't work for much anymore because it's so outdated. However, since this is the rare case of a really old aax file, it might work for you.

**Comment by rmcrackan** [comment]
also -- sorry for the delay. I've been out of town for almost 2 weeks. I'm back and am slowly working my way through this huge backlog.

**Comment by CharlieRussel** [comment]
Ok, I’m not a developer, but a long time scripter who often had to deal with others’ inconsistencies. What I would have done is:
If (multipart) AND (! Download) try (child download).

Sent from Outlook<https://aka.ms/qtex0l> on my phone. Please excuse any terseness or Autocorrect typos.
________________________________
